FirebaseMLModelDownloader/Sources/ModelInfo.swift

```diff
@@ -16,44 +16,58 @@ import Foundation
 import FirebaseCore
 
 /// Model info object with details about pending or downloaded model.
-class ModelInfo: NSObject {
+struct ModelInfo {
   /// Model name.
-  var name: String
+  let name: String
 
   // TODO: revisit UserDefaultsBacked
   /// Download URL for the model file, as returned by server.
-  @UserDefaultsBacked var downloadURL: String
+  let downloadURL: String
 
   /// Hash of the model, as returned by server.
-  @UserDefaultsBacked var modelHash: String
+  let modelHash: String
 
   /// Size of the model, as returned by server.
-  @UserDefaultsBacked var size: Int
+  let size: Int
 
   /// Local path of the model.
-  @UserDefaultsBacked var path: String?
+  var path: String?
 
   /// Initialize model info and create user default keys.
-  init(app: FirebaseApp, name: String, defaults: UserDefaults) {
+  init(name: String, downloadURL: String, modelHash: String, size: Int) {
     self.name = name
+    self.downloadURL = downloadURL
+    self.modelHash = modelHash
+    self.size = size
+  }
+
+  init?(name: String, app: FirebaseApp, defaults: UserDefaults) {
     let bundleID = Bundle.main.bundleIdentifier ?? ""
     let defaultsPrefix = "\(bundleID).\(app.name).\(name)"
-    _downloadURL = UserDefaultsBacked(
-      key: "\(defaultsPrefix).model-download-url",
-      storage: defaults
-    )
-    _modelHash = UserDefaultsBacked(key: "\(defaultsPrefix).model-hash", storage: defaults)
-    _size = UserDefaultsBacked(key: "\(defaultsPrefix).model-size", storage: defaults)
-    _path = UserDefaultsBacked(key: "\(defaultsPrefix).model-path", storage: defaults)
+    guard let downloadURL = defaults
+      .value(forKey: "\(defaultsPrefix).model-download-url") as? String,
+      let modelHash = defaults.value(forKey: "\(defaultsPrefix).model-hash") as? String,
+      let size = defaults.value(forKey: "\(defaultsPrefix).model-size") as? Int,
+      let path = defaults.value(forKey: "\(defaultsPrefix).model-path") as? String else {
+      return nil
+    }
+    self.name = name
+    self.downloadURL = downloadURL
+    self.modelHash = modelHash
+    self.size = size
+    self.path = path
   }
-}
 
-/// Named user defaults for FirebaseML.
-extension UserDefaults {
-  static var firebaseMLDefaults: UserDefaults {
-    let suiteName = "com.google.firebase.ml"
-    // TODO: reconsider force unwrapping
-    let defaults = UserDefaults(suiteName: suiteName)!
-    return defaults
+  func writeToDefaults(app: FirebaseApp, defaults: UserDefaults) throws {
```

Comment:
I don't have too strong opinion here but I would prefer to have a separate class responsible for persisting the models with public methods like `readModel(name:...)`, `save(model:)`. This approach has the following advantages:
- the storage will have the single responsibility and the model will be a pure data class
- implementation of the storage will be easy to replace or fake in the tests

This doesn't look like a blocker for the PR. We may adopt this approach later.

@ryanwilson Do you have an opinion on it?

Reply:
Hm, I'm not sure. I don't think we want to publicly expose any functionality other than a fully built model (this will be a `CustomModel` instance with read-only properties) and a `ModelDownloader` instance to fetch a model from local storage or the server given the model name. 

The only persistence we're doing at this point is model metadata, and that is primarily to check if we already have the latest model on device. But it could be worth considering if we want to refactor model info storage somehow.

Reply:
Sorry for the confusion with API term. By API I mean just the set of methods of the class, not the public SDK API.

So basically my suggestion is around slight restructuring the code to improve readability, maintainability and testability.